onnxruntime/core/framework/tensorprotoutils.cc

```diff
@@ -658,6 +670,11 @@ Status TensorProtoToTensor(const Env& env, const ORTCHAR_T* model_path,
     return Status(common::ONNXRUNTIME, common::INVALID_ARGUMENT, "string tensor can not have raw data");
   }
 
+  // NB: exit early when data is external, avoid calling UnpackTensor (which copies).
+  if (endian::native == endian::little && raw_data != nullptr && callback.f != nullptr && utils::HasExternalData(tensor_proto)) {
+    return Status::OK();
+  }
+
```

Comment:
If I understand it correctly we only want to exit early if we're loading external data that is mmap'd. If that is correct, this early exit `if` should be immediately after the GetFileContent call above so that it's inside the branch for loading external data. 

I would also expect the construction of the tensor from the raw_data that is currently lines 652-656 should be inside the early exit check. In all other cases we want to fall through and use UnpackTensor don't we?

Would also be good to document the check that callback.f is not nullptr. This is to validate there's a deleter for the raw_data right? Understanding that requires a lot of implicit knowledge without a comment explaining it.

Reply:
hi, thanks for the detailed comment.
So do you think we should continue with the single `TensorProtoToTensor` with early exit for external data; or should we revert back to 2 (`ExtDataTensorProtoToTensor` and `TensorProtoToTensor`) per your comment above?

Reply:
I would revert to having ExtDataTensorProtoToTensor and TensorProtoToTensor and a shared helper implementing the load of external data as that keeps the code simpler. 

If there's a single TensorProtoToTensor you have to allow calling with and without a pre-allocated tensor which requires a few different validations which depend on implicit things like whether the data is external. It also complicates the implementation by adding an early exit in one scenario, and the early exit condition is a little obtuse.